runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkStatefulDoFnFunction.java

```diff
@@ -186,6 +201,12 @@ public TimerInternals timerInternals() {
 
     fireEligibleTimers(key, timerInternals, doFnRunner);
 
+    if (usesOnWindowExpiration) {
+      for (BoundedWindow window : windowsSeen) {
```

Comment:
My only concern here would be that all the windows for a single key need to fit in-memory. For example year worth of 1m tumbling windows would be ~500k windows. This could be IMO quite common for batch case.

What about doing a simple optimization here by reusing the timestamp sorting mechanism that we already have in-place for the `@RequireTimeSortedInput` [1]?

Then we could simply check for expiring windows after each element / every time the timestamp progresses.

[1] https://github.com/apache/beam/blob/v2.34.0/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTransformTranslators.java#L702

Reply:
Not a bad idea, but this also holds true for timers then - in fact OnWindowExpiration can be thought of as a special type of. timer (and similar to timers, can only be used in a StatefulDoFn). If we fix one this way, we should modify both - so IMO that is best done in a separate PR.

Reply:
True, we should fix it for both. 👍 Can you please create a follow-up issue for that?

Reply:
filed BEAM-13452